src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -753,12 +806,15 @@ private void ReturnReadBufferIfEmpty()
             }
         }
 
-        private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer)
+        private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer, bool renegotiation = false)
```

Comment:
Adding this flag seems like we're abusing this method, since it skips a ton of the logic involved here. Can we just make this a separate method instead of a flag? What logic from this method do we actually need here?

Reply:
yes, it may be seen as abuse. The Windows recommendation is to call the AcceptSecurityContext with updated setting and then read from the stream. We would need to have the frame reader, call decrypt and if the decrypt fails with `Renegotiate` status we would fallow the renegotiate path that leads to `ForceAuthenticateAsync` and back. 

One option would be to make more radical cut and possibly combine it with some the refactoring. Or do the refactoring as follow-up so at least we have API and functionality for integration.

Reply:
I'm fine to defer the refactoring until a follow up PR as long as we make sure the follow up PR happens.